File: src/main/java/arcana/common/packets/PacketHandler.java
package arcana.common.packets;

import net.minecraft.network.PacketBuffer;
import net.minecraftforge.fml.network.NetworkEvent;

import java.util.function.Supplier;

public abstract class PacketHandler<T extends PacketHandler> {
    public abstract T decode(PacketBuffer buf);
    public abstract void encode(T p, PacketBuffer buf);
    public abstract void innerHandle(T packet, Supplier<NetworkEvent.Context> ctx);

    public void handle(T packet, Supplier<NetworkEvent.Context> ctx){
        ctx.get().enqueueWork(() -> innerHandle(packet, ctx));
        ctx.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/arcana/common/packets/PacketHandler.kt
package arcana.common.packets

import net.minecraft.network.PacketBuffer
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

abstract class PacketHandler<SELF : PacketHandler<SELF>> {
    abstract fun decode(buf: PacketBuffer): SELF
    abstract fun encode(packet: SELF, buf: PacketBuffer)
    abstract fun innerHandle(packet: SELF, ctx: Supplier<NetworkEvent.Context>)
    fun handle(packet: SELF, ctx: Supplier<NetworkEvent.Context>) {
        ctx.get().enqueueWork { innerHandle(packet, ctx) }
        ctx.get().packetHandled = true
    }
}